internal/consumer/service.go

```diff
@@ -0,0 +1,80 @@
+package consumer
+
+import (
+	"log"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+type Item struct {
+	By          string `bson:"by"`
+	Title       string `bson:"title"`
+	Url         string `bson:"url"`
+	Text        string `bson:"text"`
+	ItemType    string `bson:"itemType" json:"type"`
+	Descendants int    `bson:"descendants"`
+	Id          int    `bson:"id"`
+	Score       int    `bson:"score"`
+	Time        int    `bson:"time"`
+	Parent      int    `bson:"parent"`
+	Poll        int    `bson:"poll"`
+	Kids        []int  `bson:"kids"`
+	Parts       []int  `bson:"parts"`
+	Deleted     bool   `bson:"deleted"`
+	Dead        bool   `bson:"dead"`
+}
+
+// Execute - Entry point for consumer service
+func Execute(dbRepo DbRepository, dataService DataService) {
+	ids, err := dataService.getTopStories()
+	if err != nil {
+		log.Fatal(errors.Wrap(err, "Consume: "))
+	}
+
+	const workerCount = 10
+
+	idChan := make(chan int)
+	itemChan := make(chan Item)
+
+	go populateIdChan(idChan, ids)
```

Comment:
Does this need to be a go function?

Reply:
Channels is a relatively new concept for me but it looks like if i remove the GO keyword it doesnt work. Potentially something to do with a channel needing to be set up between 2 go routines?

Reply:
Your `populateIdChan` function is attempting to write to a [channel](https://github.com/JakeHumphries/gymshark-hacker-news/pull/1/files?authenticity_token=rjIYMsk5UfnEkboE6xYvbbY6Zk6%2FWIhjO%2BHympp8f5XPMzorVbirnPQTxby8sge6U3kzlCBjnPBitt3ZN%2Bb8eA%3D%3D&file-filters%5B%5D=.example&file-filters%5B%5D=.go&file-filters%5B%5D=.md&file-filters%5B%5D=dotfile#diff-6c00c19a3da7d27b57a003f96a6d9c1925cee5aa1ea535071f57b4ae3ee227f7R45) but nothing is reading from it so it blocks.

If you reorder your `Execute` function so that your workers are running (i.e. attempting to read from the channel) before you populate the `idChan` then your program will execute correctly without executing `populateIdChan` in a goroutine.

Reply:
ah makes sense. I couldn't figure out why this was only working when populateIdChan was in a go routine. Doesn't make sense to split the logic into populateIdChan now.